src/ifcconvert/JsonSerializer.cpp

```diff
@@ -0,0 +1,695 @@
+            emitted_materials.insert(mat);
+
+            // Write IfcMaterialLayerSetUsage / IfcMaterialLayerSet
+            if (mat->as<IfcMaterialLayerSetUsage>() || mat->as<IfcMaterialLayerSet>()) {
+                IfcMaterialLayerSet* layerset = mat->as<IfcMaterialLayerSet>();
+                std::string materialJsonKey = "IfcMaterialLayerSet";
+
+                if (!layerset) {
+                    materialJsonKey = "IfcMaterialLayerSetUsage";
+                    layerset = mat->as<IfcMaterialLayerSetUsage>()->ForLayerSet();
+                }
+
+                json::reference materialObject = getEmptyObjectReferenceInArray(materialJsonKey, materials);
+                materialObject["@id"] = qualify_unrooted_instance(mat);
+
+                if (layerset->hasLayerSetName()) {
+                    materialObject["@LayerSetName"] = layerset->LayerSetName();
+                }
+
+                IfcMaterialLayer::list::ptr ls = layerset->MaterialLayers();
+
+                for (IfcMaterialLayer::list::it jt = ls->begin(); jt != ls->end(); ++jt) {
+                    json::reference subMaterialObject = getEmptyObjectReferenceInArray(Type::ToString((*jt)->type()), materialObject);
+                    
+                    if ((*jt)->hasMaterial()) {
+                        subMaterialObject["@Name"] = (*jt)->Material()->Name();
+                    }
+                    
+                    format_entity_instance(*jt, subMaterialObject);
+                }
+
+                format_entity_instance((IfcUtil::IfcBaseEntity*) mat, materialObject);
+            }
+            // write IfcMaterialList
+            else if (mat->as<IfcMaterialList>()) {
+                IfcMaterial::list::ptr mats = mat->as<IfcMaterialList>()->Materials();
+
+                json::reference materialObject = getEmptyObjectReferenceInArray("IfcMaterialList", materials);
+                materialObject["@id"] = qualify_unrooted_instance(mat);
+
+                for (IfcMaterial::list::it jt = mats->begin(); jt != mats->end(); ++jt) {
+                    json::reference subMaterialObject = getEmptyObjectReferenceInArray(Type::ToString((*jt)->type()), materialObject);
+
+                    format_entity_instance(*jt, subMaterialObject);
+                }
+
+                format_entity_instance((IfcUtil::IfcBaseEntity*) mat, materialObject);
+            }
+            // write IfcMaterial
+            else if (mat->as<IfcMaterial>()) {
+                json::reference materialObject = getEmptyObjectReferenceInArray("IfcMaterial", materials);
+                materialObject["@id"] = qualify_unrooted_instance(mat);
+
+                format_entity_instance((IfcUtil::IfcBaseEntity*) mat, materialObject);
+            }
+        }
+    }
+
+	std::ofstream f(IfcUtil::path::from_utf8(json_filename).c_str());
+
+	// Write prettified json to stream
+    f << std::setw(4) << jsonRoot << std::endl;
```

Comment:
For production use prettifying the result is not needed

Reply:
Prettifying removed.